File: solutions/python/pig-latin/1/pig_latin.py
import re

def piggy_word(word):
    vowel_pattern = re.compile('^([aeiou]|yt|xr)(\w*)')
    consonant_patterns = [ re.compile('^([^aeiou]*qu)(\w*)'),
                           re.compile('^([^aeiou]+)(y\w*)'), 
                           re.compile('^([^aeiou]+)(\w*)') ]
    if vowel_pattern.match(word):
        return word+'ay'
    for pat in consonant_patterns:
        m = pat.match(word)
        if m:
            return m.groups()[-1] + m.groups()[0] + 'ay'
    

def translate(text):
    return " ".join([piggy_word(word) for word in text.split()]) 

File: solutions/dart/pig-latin/1/lib/pig_latin.dart
    String translate(String english){
     String pig = '';
     List<String> words = 
       RegExp(r"\w+").allMatches(english)
        .map((m) => m.group(0)).nonNulls.toList();
     for(var w in words){
        pig += piggifyWord(w) + ' ';
      }
      if(pig[pig.length-1]==' ')
        pig = pig.substring(0,pig.length-1);
      return pig;
    }

    String piggifyWord(String word){
      var vowels = RegExp(r"(^([aeiou]|yt|xr)(\w*))");
      List<String> consonantPatterns = [
            r"^(?<move>sch|thr|[^aeiou]qu|ch|qu|rh|th|[^aeiuo])(?<keep>\w*)"];
      // r"((?<move>[^aeiou]*[q][u])(?<keep>w*))",
           // r"((?<move>[th|^aeiouy]+)(?<keep>y\w*))", 
           // r"((?<move>[^aeiou]+)(?<keep>[aeiou]\w*))"];
      if(vowels.hasMatch(word))
        return word + 'ay';
      for(var pat in consonantPatterns){
        var regE = RegExp(pat);
        if(regE.hasMatch(word))
        {
          var match = regE.firstMatch(word)!;
           var toEnd = match.namedGroup('move')!;
           var remains = match.namedGroup('keep')!; 
          return  remains + toEnd  + 'ay';
        }
      }
      return word;
    }
  

